crypto/ex_data.c

```diff
@@ -287,7 +287,7 @@ int CRYPTO_dup_ex_data(int class_index, CRYPTO_EX_DATA *to,
         CRYPTOerr(CRYPTO_F_CRYPTO_DUP_EX_DATA, ERR_R_MALLOC_FAILURE);
         return 0;
     }
-    if (!CRYPTO_set_ex_data(to, mx - 1, NULL))
+    if (!CRYPTO_set_ex_data(to, mx - 1, CRYPTO_get_ex_data(to, mx - 1)))
```

Comment:
By the way, I might suggest a comment here since it's non-obvious what's going on. (Specifically that you need to guarantee `CRYPTO_set_ex_data` can't fail because otherwise you have no way to free `ptr`.)